app/src/org/commcare/network/HttpRequestGenerator.java

```diff
@@ -86,32 +83,37 @@ public HttpRequestGenerator(User user) {
         this(user.getUsername(), user.getCachedPwd(), user.getUserType());
     }
 
-
     public HttpRequestGenerator(String username, String password) {
         this(username, password, null);
     }
 
-    public HttpRequestGenerator(String username, String password, String userType) {
-        String domainedUsername = username;
-
-        SharedPreferences prefs = CommCareApplication._().getCurrentApp().getAppPreferences();
-
-        //TODO: We do this in a lot of places, we should wrap it somewhere
-        if (prefs.contains(USER_DOMAIN_SUFFIX)) {
-            domainedUsername += "@" + prefs.getString(USER_DOMAIN_SUFFIX, null);
-        }
+    private HttpRequestGenerator(String username, String password, String userType) {
+        String domainedUsername = buildDomainUser(username);
+        this.password = password;
+        this.userType = userType;
 
-        if (!User.TYPE_DEMO.equals(userType)) {
+        if (username != null && !User.TYPE_DEMO.equals(userType)) {
             this.credentials = new UsernamePasswordCredentials(domainedUsername, password);
-            passwordAuthentication = new PasswordAuthentication(domainedUsername, password.toCharArray());
             this.username = username;
+        } else {
+            this.credentials = null;
+            this.username = null;
         }
+    }
 
-        this.userType = userType;
+    protected static String buildDomainUser(String username) {
+        if (username != null) {
+            SharedPreferences prefs = CommCareApplication._().getCurrentApp().getAppPreferences();
+
+            if (prefs.contains(USER_DOMAIN_SUFFIX)) {
+                username += "@" + prefs.getString(USER_DOMAIN_SUFFIX, null);

```

Comment:
Shouldn't we error hard here if we don't have a domain rather than do test@null.commcarehq.org and fail later?